#!/usr/bin/env python3

# Copyright (c) 2024 PAL Robotics S.L. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import em
import sys
import os

from tiago_pro_description.tiago_pro_launch_utils import get_tiago_pro_hw_suffix

em_file_path = sys.argv[1]

no_em_extension_path = os.path.splitext(em_file_path)[0]
no_extension_path = os.path.splitext(no_em_extension_path)[0]
extension = os.path.splitext(no_em_extension_path)[1]

end_effectors = ["pal-pro-gripper", "custom", "no-end-effector", "allegro-hand"]
arm_types = ['tiago-pro', 'no-arm']
has_teleop_arms = [True, False]

for end_effector_left in end_effectors:
    for end_effector_right in end_effectors:
        for arm_type_left in arm_types:
            for arm_type_right in arm_types:

                cfg = {
                    "has_arm_left": arm_type_left != 'no-arm',
                    "has_arm_right": arm_type_right != 'no-arm',
                    "end_effector_left": end_effector_left,
                    "end_effector_right": end_effector_right,
                    "has_teleop_arms": has_teleop_arms,
                }

                with open(em_file_path, "r") as f:
                    expanded_contents = em.expand(f.read(), cfg)

                suffix = get_tiago_pro_hw_suffix(
                    arm_right=arm_type_right,
                    arm_left=arm_type_left,
                    end_effector_right=cfg["end_effector_right"],
                    end_effector_left=cfg["end_effector_left"])
                if cfg["has_teleop_arms"]:
                    suffix += "_teleop-arms"
                expanded_file_name = no_extension_path + suffix + extension
                with open(expanded_file_name, "w") as f:
                    msg = "Autogenerated file, don't edit this, edit {} instead".format(
                        os.path.basename(em_file_path))
                    if extension == ".yaml":
                        f.write("#" + msg + "\n")
                    elif extension in [".xacro", ".xml", ".srdf"]:
                        f.write("<!-- " + msg + "-->\n")

                    f.write(expanded_contents)
                print("Generated " + expanded_file_name)
